packages/jsii-diff/lib/classes-ifaces.ts

```diff
@@ -0,0 +1,203 @@
+  }
+
+  if (subclassableType(original)) {
+    // Similar as with classes, implementors should not bear the burden of having
+    // to add new mandatory fields. Only do this on types that have been explicitly
+    // marked for subclassing.
+    checkNoNewMembers(original, updated, context);
+  }
+}
+
+function checkNoNewMembers<T extends reflect.ClassType | reflect.InterfaceType>(original: T, updated: T, context: ComparisonContext) {
+    const absMemberNames = new Set(updated.members.filter(m => m.abstract).map(m => m.name));
+    const originalMemberNames = new Set(original.members.map(m => m.name));
+    for (const name of absMemberNames) {
+      if (!originalMemberNames.has(name)) {
+        context.mismatches.report(original, `adds requirement for subclasses to implement '${name}'.`);
+      }
+    }
+}
+
+function interfaceType(dataType: boolean) {
+  return dataType ? 'data interface' : 'behavior interface';
+}
+
+function describeTypeMatchingFailure(origType: reflect.TypeReference, updatedType: reflect.TypeReference, analysis: FailedAnalysis) {
+  if (origType.toString() !== updatedType.toString()) {
+    return `${updatedType} (formerly ${origType}): ${analysis.reasons.join(', ')}`;
+  } else {
+    return `${updatedType}: ${analysis.reasons.join(', ')}`;
+  }
+}
+
+function compareMethod(origClass: reflect.Type, original: reflect.Method, updated: reflect.Method, context: ComparisonContext) {
+  if (original.static !== updated.static) {
+    // tslint:disable-next-line:max-line-length
+    context.mismatches.report(origClass, `method ${original.name} was ${original.static ? 'static' : 'not static'}, is now ${updated.static ? 'static' : 'not static'}.`);
+  }
+
+  const ana = isStrengtheningPostCondition(original.returns, updated.returns);
+  if (!ana.success) {
+    // tslint:disable-next-line:max-line-length
+    context.mismatches.report(origClass, `method ${original.name}, returns ${describeTypeMatchingFailure(original.returns, updated.returns, ana)}`);
+  }
+
+  const updatedParams = updated.parameters;
+  original.parameters.forEach((param, i) => {
+    // Find the matching parameter
+    const updatedParam = findParam(updatedParams, i);
+    if (updatedParam === undefined) {
+      context.mismatches.report(origClass, `method ${original.name} argument ${param.name}, not accepted anymore.`);
+      return;
+    }
+
+    const mana = isWeakeningPreCondition(param.type, updatedParam.type);
+    if (!mana.success) {
+      // tslint:disable-next-line:max-line-length
+      context.mismatches.report(origClass, `method ${original.name} argument ${param.name}, takes ${describeTypeMatchingFailure(param.type, updatedParam.type, mana)}`);
+      return;
+    }
+  });
+}
+
+/**
+ * Check if a class/interface has been marked as @subclassable
+ */
+function subclassableType(x: reflect.Documentable) {
+  return x.docs.customTag('subclassable') !== undefined;
```

Comment:
I wonder if this should be treated as a custom tag. If we formalize this it shouldn’t be custom